File: adapters/python/blabla_adapter.py
import json
import sys

LINE_LIMIT = 1024 * 1024


class ProtocolError(Exception):
    pass


class Args:
    def __init__(self, values):
        self.values = values

    def __len__(self):
        return len(self.values)

    def string(self, index):
        return self._typed(index, str, "a string")

    def integer(self, index):
        value = self._typed(index, int, "a whole number")
        if isinstance(value, bool):
            raise ProtocolError(f"argument {index} is not a whole number")
        return value

    def boolean(self, index):
        return self._typed(index, bool, "a boolean")

    def _typed(self, index, kind, described):
        if index >= len(self.values):
            raise ProtocolError(f"argument {index} is missing")
        value = self.values[index]
        if not isinstance(value, kind):
            raise ProtocolError(f"argument {index} is not {described}")
        return value


class Adapter:
    def __init__(self, reset, observe):
        self._reset = reset
        self._observe = observe
        self._actions = {}

    def action(self, name, arity, handler):
        self._actions[name] = (arity, handler)
        return self

    def call(self, name, values):
        entry = self._actions.get(name)
        if entry is None:
            raise ProtocolError(f"unknown action: {name}")
        arity, handler = entry
        if len(values) != arity:
            raise ProtocolError(f"{name} takes {arity} arguments, got {len(values)}")
        handler(Args(values))

    def handle(self, request):
        operation = request.get("op")
        if operation == "reset":
            self._reset()
            return {"ok": True}
        if operation == "observe":
            return self._observe()
        if operation == "call":
            name = request.get("name")
            if not isinstance(name, str):
                raise ProtocolError("call has no action name")
            values = request.get("args", [])
            if not isinstance(values, list):
                raise ProtocolError("call arguments are not a list")
            self.call(name, values)
            return {"ok": True}
        raise ProtocolError(f"unknown op: {operation}")

    def serve(self, source=None, sink=None, logs=None):
        source = sys.stdin if source is None else source
        sink = sys.stdout if sink is None else sink
        logs = sys.stderr if logs is None else logs
        for line in source:
            if not line.strip():
                continue
            if len(line) > LINE_LIMIT:
                print("request exceeded the line limit", file=logs, flush=True)
                continue
            try:
                request = json.loads(line)
            except ValueError as failure:
                print(f"unreadable request: {failure}", file=logs, flush=True)
                continue
            if not isinstance(request, dict):
                print("request is not an object", file=logs, flush=True)
                continue
            try:
                result = self.handle(request)
            except ProtocolError as failure:
                result = {"ok": False, "error": str(failure)}
            sink.write(json.dumps({"id": request.get("id"), "result": result}) + "\n")
            sink.flush()
        return 0

File: adapters/typescript/blabla_adapter.ts
import { createInterface } from "node:readline";

export const LINE_LIMIT = 1024 * 1024;

export class ProtocolError extends Error {}

export class Args {
    private readonly values: unknown[];

    constructor(values: unknown[]) {
        this.values = values;
    }

    get length(): number {
        return this.values.length;
    }

    string(index: number): string {
        const value = this.at(index);
        if (typeof value !== "string") {
            throw new ProtocolError(`argument ${index} is not a string`);
        }
        return value;
    }

    integer(index: number): number {
        const value = this.at(index);
        if (typeof value !== "number" || !Number.isInteger(value)) {
            throw new ProtocolError(`argument ${index} is not a whole number`);
        }
        return value;
    }

    boolean(index: number): boolean {
        const value = this.at(index);
        if (typeof value !== "boolean") {
            throw new ProtocolError(`argument ${index} is not a boolean`);
        }
        return value;
    }

    private at(index: number): unknown {
        if (index >= this.values.length) {
            throw new ProtocolError(`argument ${index} is missing`);
        }
        return this.values[index];
    }
}

export type Handler = (args: Args) => void;

export class Adapter {
    private readonly actions = new Map<string, { arity: number; handler: Handler }>();
    private readonly onReset: () => void;
    private readonly onObserve: () => unknown;

    constructor(onReset: () => void, onObserve: () => unknown) {
        this.onReset = onReset;
        this.onObserve = onObserve;
    }

    action(name: string, arity: number, handler: Handler): Adapter {
        this.actions.set(name, { arity, handler });
        return this;
    }

    call(name: string, values: unknown[]): void {
        const entry = this.actions.get(name);
        if (entry === undefined) {
            throw new ProtocolError(`unknown action: ${name}`);
        }
        if (values.length !== entry.arity) {
            throw new ProtocolError(
                `${name} takes ${entry.arity} arguments, got ${values.length}`,
            );
        }
        entry.handler(new Args(values));
    }

    handle(request: Record<string, unknown>): unknown {
        switch (request.op) {
            case "reset":
                this.onReset();
                return { ok: true };
            case "observe":
                return this.onObserve();
            case "call": {
                const name = request.name;
                if (typeof name !== "string") {
                    throw new ProtocolError("call has no action name");
                }
                const values = request.args ?? [];
                if (!Array.isArray(values)) {
                    throw new ProtocolError("call arguments are not a list");
                }
                this.call(name, values);
                return { ok: true };
            }
            default:
                throw new ProtocolError(`unknown op: ${String(request.op)}`);
        }
    }

    serve(source: NodeJS.ReadableStream = process.stdin): void {
        const lines = createInterface({ input: source });
        lines.on("line", (line) => {
            if (line.trim() === "") {
                return;
            }
            if (line.length > LINE_LIMIT) {
                process.stderr.write("request exceeded the line limit\n");
                return;
            }
            let request: unknown;
            try {
                request = JSON.parse(line);
            } catch (failure) {
                process.stderr.write(`unreadable request: ${String(failure)}\n`);
                return;
            }
            if (typeof request !== "object" || request === null || Array.isArray(request)) {
                process.stderr.write("request is not an object\n");
                return;
            }
            const envelope = request as Record<string, unknown>;
            let result: unknown;
            try {
                result = this.handle(envelope);
            } catch (failure) {
                if (!(failure instanceof ProtocolError)) {
                    throw failure;
                }
                result = { ok: false, error: failure.message };
            }
            process.stdout.write(`${JSON.stringify({ id: envelope.id ?? null, result })}\n`);
        });
    }
}
